### DISCUSSION TITLE
Dynamic where clauses

### DISCUSSION THREAD

**Discussion by mhpaler** [graphprotocol/graph-client#282] [Q&A] [answered]
I'm only first level brown belt in Subgraph/GraphQL so not even sure a feature like this makes sense for the graph-client, but here goes.

I'm finding myself writing duplicate queries to handle different `where` clauses, and obviously this feels redundant and clunky. Doing a bit of research I came across this interesting article which describes a generic way to handle passing different `filter` clauses into a single query from a node based GraphQL api client.

https://dev.to/mgustus/filtering-graphql-query-using-typescript-and-typeorm-2l49

Curious if this, or something like it, would make sense for the graph-client?

**Comment by mhpaler** [comment]
In context to this, question, can we do something like this:

```
query get_items($where: SomeOtherType) {
  items(where: $where) {
    name
  }
}
```
Where `SomeOtherType` could have any shape we want? If so, where & how would one define `SomeOtherType`?

**Comment by dotansimha** [comment, marked as answer]
Hi @mpaler ! 
You can leverage the concept of GraphQL `variables` to achieve that. 
First, you need to find the actual GraphQL `input` type that's defined for the `where` argument you with to match, and then use that `input` type for your GraphQL query, as a variable (similar to the example you shared). You can do that with GraphiQL, or just introspect the schema and check the API docs of your field.

For example, [for this Subgraph](https://api.thegraph.com/subgraphs/name/sushiswap/exchange), you can see in the docs that the `User_filter` input type is used for the `where`:
<img width="355" alt="image" src="https://user-images.githubusercontent.com/3680083/194738111-26746358-c75f-4612-b072-77f065a272b5.png">

So you can write a query that expects this `input` type as a variable, and then match it to the `where` argument

```graphql
query filteredUsers($usersFilter: User_filter) { # define the variable
  users(where: $usersFilter) { # pair it with the argument
    id
    liquidityPositions {
      id
      pair {
        token0 {
          id
        }
        token1 {
          id
        }
      }
    }
  }
} 
```

Now, when you run your GraphQL query, you can use this single GraphQL query, but with different (=dynamic) variables for each filter that you might need:

<img width="1147" alt="image" src="https://user-images.githubusercontent.com/3680083/194738173-ce8c7e7a-e264-4d56-b3af-238e11711e17.png">

> This is the best practice in GraphQL, also because the query remains static, and only the value of `variables` changes every time.

**Reply by mhpaler** [reply]
@dotansimha thank you!! Been going in circles for two days trying to figure out how I would adjust my schema so I could pass in a filter argument of any shape. This completely makes sense and solves it!!
